pkg/querier/stats_renderer_test.go

```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querier
+
+import (
+	"context"
+	"net/http"
+	"net/http/httptest"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/grafana/dskit/user"
+	"github.com/grafana/regexp"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/common/promslog"
+	"github.com/prometheus/common/route"
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/storage"
+	v1 "github.com/prometheus/prometheus/web/api/v1"
+	"github.com/stretchr/testify/require"
+
+	mimir_stats "github.com/grafana/mimir/pkg/querier/stats"
+	"github.com/grafana/mimir/pkg/storage/series"
+)
+
+func TestStatsRenderer(t *testing.T) {
+
+	testCases := []struct {
+		name    string
+		samples uint64
+	}{
+		{name: "ZeroSamples", samples: 0},
+		{name: "OneSample", samples: 1},
+		{name: "TenSamples", samples: 10},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			engine := promql.NewEngine(promql.EngineOpts{
+				Logger:     nil,
+				MaxSamples: 100,
+				Timeout:    5 * time.Second,
+			})
+
+			now := model.Time(time.Now().UnixMilli())
+
+			var seriesList []storage.Series
+
+			for i := 0; i < int(tc.samples); i++ {
+				seriesList = append(seriesList, series.NewConcreteSeries(
+					labels.FromStrings("__name__", "test", "serial", strconv.Itoa(i)),
+					[]model.SamplePair{{Timestamp: now, Value: model.SampleValue(i)}}, nil,
+				))
+			}
+
+			q := &mockSampleAndChunkQueryable{
+				queryableFn: func(_, _ int64) (storage.Querier, error) {
+					return &mockQuerier{
+						selectFn: func(_ context.Context, _ bool, _ *storage.SelectHints, _ ...*labels.Matcher) storage.SeriesSet {
+							return series.NewConcreteSeriesSetFromUnsortedSeries(seriesList)
+						},
+					}, nil
+				},
+			}
```

Comment:
[nit] Something that could simplify this: instead of creating a mock queryable, use a `promqltest.LoadedStorage` with ten series, and use different expressions for each test case to limit the number of series (and therefore samples) selected.

This would also allow you to move most of the setup code out of the `t.Run`, as you could share the same queryable and API. This would make the `t.Run` more focused on the specific test case.

Reply:
Thank you for the idea 😄.